### PR TITLE
Add Smart Filters

### PR DESCRIPTION
**Pull request by krossgg**
## What
Adds Smart Filters as a new Filter type
They are made from Ruby.

## Implementation Details
Implemented similarly to 1.12.
Enum subclass SmartFilteringMode that dictates which machine mode is being filtered by the current filter.
Each Mode has a cache map of ItemStack -> int to alleviate computation time.

Selectively filter those items just like Orwell would've wanted

### CODE REVIEW

**Comment by YoungOnionMC** [comment]
gorg orwel

File: src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java [hunk 1]
```diff
@@ -1997,6 +1994,12 @@ public Component getItemName(ItemStack stack) {
                     new CoverPlaceBehavior(GTCovers.ITEM_FILTER)))
             .onRegister(materialInfo(new ItemMaterialInfo(new MaterialStack(GTMaterials.Zinc, GTValues.M * 2))))
             .onRegister(compassNode(GTCompassSections.COVERS, GTCompassNodes.COVER)).register();
+    public static ItemEntry<ComponentItem> SMART_ITEM_FILTER = REGISTRATE
+            .item("smart_item_filter", ComponentItem::create)
+            .lang("Smart Item Filter")
+            .onRegister(attach(new ItemFilterBehaviour(SmartItemFilter::loadFilter), new CoverPlaceBehavior(GTCovers.ITEM_FILTER)))
+            .onRegister(materialInfo(new ItemMaterialInfo(new MaterialStack(GTMaterials.Zinc, GTValues.M * 2))))
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: isn't this zinc positive since you only use 6 foil for it?

Comment (krossgg) [reply src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: i have no clue how the item builder works i just copied it from the item filter 😭

Comment (screret) [reply src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: Set it to like M*3/2

**Review by YoungOnionMC — APPROVED** [review]

**Pull request merged**